packages/react-jss/src/JssProvider.js

```diff
@@ -20,85 +20,92 @@ type Props = {
+    jss: PropTypes.instanceOf(defaultJss.constructor),
     generateId: PropTypes.func,
     classNamePrefix: PropTypes.string,
     disableStylesGeneration: PropTypes.bool,
-    children: PropTypes.node.isRequired
+    children: PropTypes.node.isRequired,
+    sheetOptions: PropTypes.shape({
+      media: PropTypes.string,
+      element: PropTypes.instanceOf(HTMLStyleElement)
+    })
   }
 
-  static childContextTypes = contextTypes
+  managers: Managers = {}
 
-  static contextTypes = contextTypes
-
-  // JssProvider can be nested. We allow to overwrite any context prop at any level.
-  // 1. Check if there is a value passed over props.
-  // 2. If value was passed, we set it on the child context.
-  // 3. If value was not passed, we proxy parent context (default context behaviour).
-  getChildContext(): Context {
+  createContext(outerContext: Context): Context {
     const {
       registry,
       classNamePrefix,
-      jss: localJss,
+      jss,
       generateId,
-      disableStylesGeneration
+      disableStylesGeneration,
+      sheetOptions
     } = this.props
-    // eslint-disable-next-line react/react/destructuring-assignment
-    const sheetOptions = this.context[ns.sheetOptions] || {}
-    const context: Context = {[ns.sheetOptions]: sheetOptions}
+    // Clone the outer context
+    const context = {
+      ...outerContext,
+      managers: this.managers
+    }
 
-    if (registry) {
-      context[ns.sheetsRegistry] = registry
-      // This way we identify a new request on the server, because user will create
-      // a new Registry instance for each.
-      if (registry !== this.registry) {
-        // We reset managers because we have to regenerate all sheets for the new request.
-        this.managers = {}
-        this.registry = registry
+    if (sheetOptions) {
+      warning(
+        'generateId' in sheetOptions,
+        '[JSS] Pass generateId as a prop instead of sheet options.'
+      )
+
+      context.sheetOptions = {
+        ...context.sheetOptions,
+        ...sheetOptions
       }
     }
 
-    // Make sure we don't loose managers when JssProvider is used inside of a stateful
-    // component which decides to rerender.
-    context[ns.managers] = this.managers
+    if (registry) {
+      context.registry = registry
+    }
 
     if (generateId) {
-      sheetOptions.generateId = generateId
-    } else if (!sheetOptions.generateId) {
-      if (!this.generateId) {
-        let createGenerateId = createGenerateIdDefault
-        if (localJss && localJss.options.createGenerateId) {
-          createGenerateId = localJss.options.createGenerateId
-        }
-        // Make sure we don't loose the generator when JssProvider is used inside of a stateful
-        // component which decides to rerender.
-        this.generateId = createGenerateId()
-      }
+      context.sheetOptions.generateId = generateId
+    } else if (!context.sheetOptions.generateId) {
+      this.generateId = createGenerateId()
```

Comment:
why do we assign generateId to the instance? we don't seem to be using it from the instance

Reply:
Fixed.